packages/notifi-web-push-service-worker/lib/notifi-service-worker.js

```diff
@@ -0,0 +1,261 @@
+  var padding = '='.repeat((4 - base64String.length % 4) % 4);
+  var base64 = (base64String + padding).replace(/-/g, '+').replace(/_/g, '/');
+
+  var rawData = self.atob(base64);
+  var outputArray = new Uint8Array(rawData.length);
+
+  for (var i = 0; i < rawData.length; ++i) {
+    outputArray[i] = rawData.charCodeAt(i);
+  }
+  return outputArray;
+}
+
+async function createWebPushTarget(subscription, vapidPublicKey) {
+  try {
+    const conv = (val) => self.btoa(String.fromCharCode.apply(null, new Uint8Array(val)));
+    const targetGroups = await client.getTargetGroups();
+    const defaultTargetGroup = targetGroups.find((targetGroup) => targetGroup.name === 'Default');
+    const webPushTargetIds = defaultTargetGroup?.webPushTargets?.map((t) => t?.id) ?? [];
+
+    const webPushTargetResponse = await client.createWebPushTarget({
+      vapidPublicKey: vapidPublicKey,
+      endpoint: subscription.endpoint,
+      auth: conv(subscription.getKey("auth")),
+      p256dh: conv(subscription.getKey("p256dh"))
+    })
+
+    if (!webPushTargetResponse.createWebPushTarget.webPushTarget || !webPushTargetResponse.createWebPushTarget.webPushTarget?.id) {
+      console.error('Failed to create web push target. CreateWebPushTargetMutation failed.');
+      return;
+    }
+
+    webPushTargetIds.push(webPushTargetResponse.createWebPushTarget.webPushTarget?.id);
+    await client.ensureTargetGroup({
+      name: 'Default',
+      emailAddress: defaultTargetGroup?.emailTargets[0]?.emailAddress,
+      phoneNumber: defaultTargetGroup?.smsTargets[0]?.phoneNumber,
+      telegramId: defaultTargetGroup?.telegramTargets[0]?.telegramId,
+      discordId: defaultTargetGroup?.discordTargets[0]?.name,
+      slackId: defaultTargetGroup?.slackChannelTargets[0]?.name,
+      walletId: defaultTargetGroup?.web3Targets[0]?.name,
+      webPushTargetIds: webPushTargetIds
+    });
+    await db.set(webPushTargetIdKey, webPushTargetResponse.createWebPushTarget.webPushTarget?.id);
+  }
+  catch (err) {
+    console.error(err, "Failed to create web push target.")
+  }
+}
+
+async function updateWebPushTarget(subscription, webPushTargetId) {
+  try {
+    const conv = (val) => self.btoa(String.fromCharCode.apply(null, new Uint8Array(val)));
+    await client.updateWebPushTarget({
+      id: webPushTargetId,
+      endpoint: subscription.endpoint,
+      auth: conv(subscription.getKey("auth")),
+      p256dh: conv(subscription.getKey("p256dh"))
+    });
+  }
+  catch (err) {
+    console.error(err, "Failed to update web push target.")
+  }
+}
+
+async function createOrUpdateWebPushTarget(subscription, vapidPublicKey) {
+  const webPushTargetId = await db.get(webPushTargetIdKey);
+  if (!webPushTargetId || webPushTargetId === '') {
+    await createWebPushTarget(subscription, vapidPublicKey);
+  }
+  else {
+    const getWebPushTargetsResponse = await client.getWebPushTargets({ ids: [webPushTargetId] });
+
+    if (getWebPushTargetsResponse.nodes.length !== 1) {
```

Comment:
Not sure if it is correct. What happens when the `length` is 2. Is it supposed to create a new target?

Reply:
It's querying specifically for the webPushTarget. Length is 2 should never happen

Reply:
I refactor a little to make it more clear at #689 FYI